styx-common/src/main/java/com/spotify/styx/util/GoogleIdTokenValidator.java

```diff
@@ -0,0 +1,218 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GoogleIdTokenValidator {
+
+  private static final Logger logger = LoggerFactory.getLogger(GoogleIdTokenValidator.class);
+
+  private static final Pattern EMAIL_PATTERN = Pattern.compile("^.+@(.+)$");
+
+  private static final Pattern SERVICE_ACCOUNT_PATTERN =
+      Pattern.compile("^.+@(.+)\\.iam\\.gserviceaccount\\.com$");
+
+  private static final long VALIDATED_EMAIL_CACHE_SIZE = 1000;
+
+  private final GoogleIdTokenVerifier googleIdTokenVerifier;
+
+  private final CloudResourceManager cloudResourceManager;
+
+  private final Iam iam;
+
+  private final Set<String> domainWhitelist;
+
+  private final Set<String> projectCache = Sets.newConcurrentHashSet();
+
+  private final Cache<String, String> validatedEmailCache = CacheBuilder.newBuilder()
+      .maximumSize(VALIDATED_EMAIL_CACHE_SIZE)
+      .build();
+
+  public GoogleIdTokenValidator(GoogleIdTokenVerifier googleIdTokenVerifier,
+                         CloudResourceManager cloudResourceManager,
+                         Iam iam,
+                         Set<String> domainWhitelist) {
+    this.googleIdTokenVerifier =
+        Objects.requireNonNull(googleIdTokenVerifier, "googleIdTokenVerifier");
+    this.cloudResourceManager =
+        Objects.requireNonNull(cloudResourceManager, "cloudResourceManager");
+    this.iam = Objects.requireNonNull(iam, "iam");
+    this.domainWhitelist = Objects.requireNonNull(domainWhitelist, "domainWhitelist");
+  }
+
+  public void cacheProjects() throws IOException {
+    final CloudResourceManager.Projects.List request = cloudResourceManager.projects().list();
+
+    ListProjectsResponse response;
+    do {
+      response = request.execute();
+      if (response.getProjects() == null) {
+        continue;
+      }
+      for (Project project : response.getProjects()) {
+        projectCache.add(project.getProjectId());
+      }
+      request.setPageToken(response.getNextPageToken());
+    } while (response.getNextPageToken() != null);
+
+    logger.info("project cache loaded");
+  }
+
+  public GoogleIdToken validate(String token) {
+    final GoogleIdToken googleIdToken;
+    try {
+      googleIdToken = verifyIdToken(token);
+    } catch (IOException e) {
+      logger.warn("failed to verify token");
+      return null;
+    }
+
+    if (googleIdToken == null) {
+      return null;
+    }
+
+    final String email = googleIdToken.getPayload().getEmail();
+
+    final Matcher matcher = EMAIL_PATTERN.matcher(email);
+    if (matcher.matches()) {
+      final String domain = matcher.group(1);
+      if (domainWhitelist.contains(domain)) {
+        logger.debug("domain {} in whitelist", domain);
+        return googleIdToken;
+      }
+    } else {
+      logger.warn("invalid email address {}", email);
+      return null;
+    }
+
+    if (validatedEmailCache.getIfPresent(email) != null) {
```

Comment:
We might need negative caching as well.

Reply:
I was worried by temporary GCP glitch caused service account validation failure. If we put that SA to negative cache, it would not be able to make request. DDoS may not very likely to happen for a service like Styx.